src/java.base/share/classes/jdk/internal/lang/stable/StableValueImpl.java

```diff
@@ -0,0 +1,419 @@
+        throw shouldNotReachHere();
+    }
+
+    @ForceInline
+    @Override
+    public V setIfUnset(V value) {
+        if (isSet() || isError()) {
+           return orThrow();
+        }
+        final var m = acquireMutex();
+        if (isMutexNotNeeded(m)) {
+            return orThrow();
+        }
+        synchronized (m) {
+            if (isSet() || isError()) {
+                return orThrow();
+            }
+            setValue(value);
+            return value;
+        }
+    }
+
+    @ForceInline
+    @Override
+    public boolean trySet(V value) {
+        if (isSet() || isError()) {
+            return false;
+        }
+        final var m = acquireMutex();
+        if (isMutexNotNeeded(m)) {
+            return false;
+        }
+        synchronized (m) {
+            if (isSet() || isError()) {
+                return false;
+            }
+            setValue(value);
+            return true;
+        }
+    }
+
+    @ForceInline
+    @Override
+    public V computeIfUnset(Supplier<? extends V> supplier) {
+        // Optimistically try plain semantics first
+        final V v = value;
+        if (v != null) {
+            // If we happen to see a non-null value under
+            // plain semantics, we know a value is set.
+            return v;
+        }
+        if (state == NULL) {
+            // If we happen to see a state value of NULL under
+            // plain semantics, we know a value is set to `null`.
+            return null;
+        }
+        // Now, fall back to volatile semantics.
+        return computeIfUnsetVolatile(supplier);
+    }
+
+    @DontInline // Slow-path taken at most once per thread if set
+    private V computeIfUnsetVolatile(Supplier<? extends V> supplier) {
+        // This is intentionally an old switch statement as it generates
+        // more compact byte code.
+        switch (stateVolatile()) {
+            case UNSET:    { return computeIfUnsetVolatile0(supplier); }
+            case NON_NULL: { return valueVolatile(); }
+            case NULL:     { return null; }
+            case ERROR:    { throw StableUtil.error(this); }
+        }
+        throw shouldNotReachHere();
+    }
+
+    private V computeIfUnsetVolatile0(Supplier<? extends V> supplier) {
+        final var m = acquireMutex();
+        if (isMutexNotNeeded(m)) {
+            return orThrow();
+        }
+        synchronized (m) {
+            // A value is already set
+            if (state != UNSET) {
+                return orThrow();
+            }
+            // A value is not set
+            if (supplying) {
+                throw stackOverflow(supplier, null);
+            }
+            try {
+                supplying = true;
+                try {
+                    V newValue = supplier.get();
+                    setValue(newValue);
+                    return newValue;
+                } catch (Throwable t) {
+                    putState(ERROR);
+                    putMutex(t.getClass());
```

Comment:
Should we cache the exception instance so we can rethrow it in future ERROR state `orThrow` calls?

Reply:
We considered recording the entire exception instance but for security reasons, we ended up just recording the type of exception. I will add a comment explaining this in the code.

Reply:
Thanks for this clarification. Makes sense.